python/ray/serve/metric/sink.py

```diff
@@ -0,0 +1,139 @@
+from typing import Dict
+from collections import Counter, namedtuple
+
+import ray
+from ray.serve.metric.types import MetricType, MetricMetadata, MetricBatch
+from ray.serve.utils import _get_logger
+
+logger = _get_logger()
+
+
+def make_metric_namedtuple(metadata: MetricMetadata, record: MetricBatch):
+    fields = ["name", "type"]
+    fields += list(metadata.default_labels.keys())
+    fields += list(record.labels.keys())
+
+    merged_labels = metadata.default_labels.copy()
+    merged_labels.update(record.labels)
+
+    tuple_type = namedtuple(metadata.name, fields)
+    return tuple_type(name=metadata.name, type=metadata.type, **merged_labels)
+
+
+class BaseSink:
+    """A sink is the place to store or forward metrics to external services"""
+
+    def __init__(self):
+        self.metadata = dict()
+        self.recording = []
+        logger.debug("Sink initialized {}".format(type(self)))
+
+    def push_batch(self, metadata: Dict[str, MetricMetadata],
+                   batch: MetricBatch):
+        self.metadata.update(metadata)
+        self.recording.extend(batch)
+        self.compact()
+
+    def compact(self):
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+    def get_metric(self):
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+
+@ray.remote(num_cpus=0)
+class InMemorySink(BaseSink):
+    def __init__(self):
+        super().__init__()
+
+        # Keep track of counters
+        self.counters: Counter[namedtuple, float] = Counter()
+        # Keep track of latest observation of measures
+        self.latest_measures: Dict[namedtuple, float] = dict()
+
+    def compact(self):
+        for record in self.recording:
```

Comment:
I'm not a huge fan of this inheritance-based pattern because it's hard to follow what exactly `compact` should do and what is done by the superclass.

How about changing this to users providing a class that implements `compact` that takes the latest recordings as input, then always clear them after calling `compact` in the caller.

Reply:
changed the signature of compact to `export(self, metric_metadata, metric_batch)`, this should give a clear idea that the entire batch should be consumed.